File: Data Structures & Algorithms/house-robber/submission-0.py
class Solution:
    def helper(self, idx: int, nums: List[int], dp: List[int]) -> int:
        if idx == 0:
            return nums[0]
        if idx < 0:
            return 0
        
        if dp[idx] != -1:
            return dp[idx]

        pick = nums[idx] + self.helper(idx-2,nums,dp)
        not_pick = self.helper(idx-1,nums,dp)

        dp[idx] = max(pick, not_pick)
        return dp[idx]

    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [-1]*(n)
        return self.helper(n-1,nums,dp)

File: Data Structures & Algorithms/combination-target-sum/submission-0.cpp
class Solution {
public:
    void helper(vector<int>& nums, int idx, vector<int>& ds, vector<vector<int>>& ans, int target){
        int n = nums.size();
        if(idx == n){
            if(target == 0){
                ans.push_back(ds);
            }
            return;
        }

        if(nums[idx] <= target){
            ds.push_back(nums[idx]);
            helper(nums,idx,ds,ans,target-nums[idx]);
            ds.pop_back();
        }
        helper(nums,idx+1,ds,ans,target);
    }
    vector<vector<int>> combinationSum(vector<int>& nums, int target) {
        vector<int> ds;
        vector<vector<int>> ans;
        helper(nums,0,ds,ans,target);
        return ans;
    }
};
